### PR TITLE
Add missing .to_sparse(ndim) gradient, fixes #46720

### LINKED ISSUE #46720: autograd for to_sparse(dim)
## 🚀 Feature
autograd for to_sparse(dim)

## Motivation
This is related to another GitHub [issue](https://github.com/pytorch/pytorch/issues/18111), that issue is fixed by the [commit](https://github.com/pytorch/pytorch/pull/20458). However, it seems it only supports non-parameter version. I want to have autograd for to_sparse(dim). 

To be more concrete, the following (case 1) is supported
```
>>> torch.randn(2, 3).requires_grad_(True).to_sparse()
tensor(indices=tensor([[0, 0, 0, 1, 1, 1],
                       [0, 1, 2, 0, 1, 2]]),
       values=tensor([ 0.3336,  0.7028, -1.0183, -0.3018, -0.6059, -1.0752]),
       size=(2, 3), nnz=6, layout=torch.sparse_coo, grad_fn=<ToSparseBackward>)
```
 the following (case 2) is not supported
```
>>> torch.randn(2, 3).requires_grad_(True).to_sparse(1)
tensor(indices=tensor([[0, 1]]),
       values=tensor([[ 1.1035,  2.0671,  0.5015],
                      [ 0.0975,  0.3090, -1.1393]]),
       size=(2, 3), nnz=2, layout=torch.sparse_coo, grad_fn=<NotImplemented>)
```

Is it possible to support case 2?

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @aocsa @mruberry

### PR DESCRIPTION
**Pull request by MingweiSamuel**
Fixes #46720

Not sure what the best way to test this is. I did run the tests with:

https://github.com/pytorch/pytorch/blob/c31ced42469cd5f0b4e758f5484be40a9515bc38/test/test_autograd.py#L5047-L5048

changed to `.to_sparse(1)`, which failed before but passes now.

### CODE REVIEW

**Comment by albanD** [comment]
You shouldn't have to do that as the formula (should) be applied to all the overloads of a function. So in particular the formula above for `to_sparse` should be enough.

**Comment by nairbv** [comment]
Regardless of what the correct fix is, we should add the relevant example to the unit test too

**Comment by kylevedder** [comment]
What is the status of this PR/other PRs for this feature?

If this is the most mature PR, @albanD can you point me to/provide a bit more background on the mechanics of the autogenerated code so I can try to answer your question? Compiling from source with this PR has been very painful, but this missing feature is a showstopper on several veins of research so I would like to get an actual fix merged so that it's not necessary in the future.

**Comment by albanD** [comment]
Hi,

Given the issue linked above, it looks like the right fix.
It would be great to have a test for it. You can add a new line next to the existing to_sparse() test here: https://github.com/pytorch/pytorch/blob/ee93a348de1bd2984a044bf9a53d4b926ed5346b/torch/testing/_internal/common_methods_invocations.py#L6504 with the extra argument.

**Comment by albanD** [comment]
This has been fixed by https://github.com/pytorch/pytorch/pull/58413
Thanks for helping!

**Comment by MingweiSamuel** [comment]
Thanks @kylevedder :)
